devito/types/dimension.py

```diff
@@ -892,6 +895,18 @@ def condition(self):
     def indirect(self):
         return self._indirect
 
+    @property
+    def fact_index(self):
```

Comment:
can you not move this whole logic into `__init_finalize__` and set index to the proper value from the get go?